File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Kadane's algorithm
        
        if len(s) == 0:
            return 0
        max_current = s[0]
        max_global = s[0]
        
        for i in range(1, len(s)):
            if s[i] in max_current:
                
                starting_index = max_current.index(s[i])
                max_current = max_current[max_current.index(s[i]) + 1: i] + s[i]
            else:
                max_current += s[i]
            
            if len(max_current) > len(max_global):
                max_global = max_current
        return len(max_global)
            
        
    
    

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    
    if len(s) == 0 {
        return 0
    }
    
    // Kadane's algorithm
    currentMax := string(s[0])
    globalMax := string(s[0])
    for i := 1; i < len(s); i++ {  
        for j := 0; j < len(currentMax); j++ {
            if s[i] == currentMax[j] {
                currentMax = currentMax[j+1:]
                break
            }
        }
        
        currentMax += string(s[i])
        
        if len(currentMax) > len(globalMax) {
            globalMax = currentMax
        }
    }
    
    return len(globalMax)
}